flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/HeapInternalTimerService.java

```diff
@@ -0,0 +1,317 @@
+	public long currentWatermark() {
+		return currentWatermark;
+	}
+
+	@Override
+	public void registerProcessingTimeTimer(N namespace, long time) {
+		InternalTimer<K, N> timer = new InternalTimer<>(time, (K) keyContext.getCurrentKey(), namespace);
+
+		// make sure we only put one timer per key into the queue
+		if (processingTimeTimers.add(timer)) {
+
+			InternalTimer<K, N> oldHead = processingTimeTimersQueue.peek();
+			long nextTriggerTime = oldHead != null ? oldHead.getTimestamp() : Long.MAX_VALUE;
+
+			processingTimeTimersQueue.add(timer);
+
+			// check if we need to re-schedule our timer to earlier
+			if (time < nextTriggerTime) {
+				if (nextTimer != null) {
+					nextTimer.cancel(false);
+				}
+				nextTimer = processingTimeService.registerTimer(time, this);
+			}
+		}
+	}
+
+	@Override
+	public void registerEventTimeTimer(N namespace, long time) {
+		InternalTimer<K, N> timer = new InternalTimer<>(time, (K) keyContext.getCurrentKey(), namespace);
+		if (watermarkTimers.add(timer)) {
+			watermarkTimersQueue.add(timer);
+		}
+	}
+
+	@Override
+	public void deleteProcessingTimeTimer(N namespace, long time) {
+		InternalTimer<K, N> timer = new InternalTimer<>(time, (K) keyContext.getCurrentKey(), namespace);
+
+		if (processingTimeTimers.remove(timer)) {
+			processingTimeTimersQueue.remove(timer);
+		}
+	}
+
+	@Override
+	public void deleteEventTimeTimer(N namespace, long time) {
+		InternalTimer<K, N> timer = new InternalTimer<>(time, (K) keyContext.getCurrentKey(), namespace);
+		if (watermarkTimers.remove(timer)) {
+			watermarkTimersQueue.remove(timer);
+		}
+	}
+
+	@Override
+	public void onProcessingTime(long time) throws Exception {
+		// null out the timer in case the Triggerable calls registerProcessingTimeTimer()
+		// inside the callback.
+		nextTimer = null;
+
+		InternalTimer<K, N> timer;
+
+		while ((timer  = processingTimeTimersQueue.peek()) != null && timer.getTimestamp() <= time) {
+
+			processingTimeTimers.remove(timer);
+			processingTimeTimersQueue.remove();
+
+			keyContext.setCurrentKey(timer.getKey());
+			triggerTarget.onProcessingTime(timer);
+		}
+
+		if (timer != null) {
+			if (nextTimer == null) {
+				nextTimer = processingTimeService.registerTimer(timer.getTimestamp(), this);
+			}
+		}
+	}
+
+	public void advanceWatermark(long time) throws Exception {
+		currentWatermark = time;
+
+		InternalTimer<K, N> timer = watermarkTimersQueue.peek();
+
+		while (timer != null && timer.getTimestamp() <= time) {
+			watermarkTimers.remove(timer);
+			watermarkTimersQueue.remove();
+
+			keyContext.setCurrentKey(timer.getKey());
+			triggerTarget.onEventTime(timer);
+
+			timer = watermarkTimersQueue.peek();
+		}
+	}
+
+	public void snapshotTimers(OutputStream outStream) throws IOException {
+		InstantiationUtil.serializeObject(outStream, keySerializer);
+		InstantiationUtil.serializeObject(outStream, namespaceSerializer);
+
+		DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(outStream);
+
```

Comment:
Just for uniformity, you could use the Sets for both processing and event time timers when iterating over them to checkpoint.